runners/flink/src/test/java/org/apache/beam/runners/flink/streaming/UnboundedSourceWrapperTest.java

```diff
@@ -300,21 +304,37 @@ public void close() {}
 
       sourceThread.start();
 
-      while (true) {
-        if (!caughtExceptions.isEmpty()) {
-          fail("Caught exception(s): " + Joiner.on(",").join(caughtExceptions));
-        }
-        if (seenWatermark.get()) {
-          break;
-        }
+      while (flinkWrapper
+          .getLocalReaders()
+          .stream()
+          .anyMatch(reader -> reader.getWatermark().getMillis() == 0)) {
+        // readers haven't been initialized
         Thread.sleep(50);
+      }
 
-        // Need to advance this so that the watermark timers in the source wrapper fire
-        // Synchronize is necessary because this can interfere with updating the PriorityQueue
-        // of the ProcessingTimeService which is also accessed through UnboundedSourceWrapper.
-        synchronized (checkpointLock) {
-          testHarness.setProcessingTime(Instant.now().getMillis());
-        }
+      // Need to advance this so that the watermark timers in the source wrapper fire
+      // Synchronize is necessary because this can interfere with updating the PriorityQueue
+      // of the ProcessingTimeService which is also accessed through UnboundedSourceWrapper.
+      synchronized (testHarness.getCheckpointLock()) {
```

Comment:
Unfortunately this is not possible. The watermark timer is not registered then so this would not to lead to a watermark emission, hence the waiting before for the readers before.

Reply:
Now I see that you changed the initial value to `Long.MIN_VALUE`. What is the significance of changing the processing time to 0 in the next line?

Reply:
The idea is to make this test more meaningful by testing emission of multiple watermarks by the source. The initial value before was arbitrary as well, as it was `System.currentTimeMillis()`.

Afterwards we advance to 0, but it could be any value between MIN_VALUE and MAX_VALUE, but it has to be larger than currentTime + autowatermark interval. This emits a watermark from the TestCoutingSource.

Then we advance it to MAX_VALUE to test the emission of the final watermark.